StringHelper: Use functional programming to improve code readability #7838

darrenwee opened the pull request:
Fixes #7838

**Outline of Solution**

<!-- Tell us how you solved the issue. -->
I refactored methods to use Java 8's streams API to improve readability. I left some functions untouched because I felt refactoring them would not improve the code quality or I wasn't sure how to refactor them to use streams. Examples of those methods are [`csvToHtmlTable`](https://github.com/darrenwee/teammates/blob/84b3b649d5ac723acf1b2d079ff9e35cd15f75bd/src/main/java/teammates/common/util/StringHelper.java#L332) and [`byteArrayToHexString`](https://github.com/darrenwee/teammates/blob/84b3b649d5ac723acf1b2d079ff9e35cd15f75bd/src/main/java/teammates/common/util/StringHelper.java#L307).

I see #8257 was also filed two hours ago. I posted in #7838 that I would take the issue. Possible to merge the two together?

(sorry for the delay, I'm on vacation and don't have consistent access to my laptop)

thenaesh commented:
@darrenwee Let me know when this is ready for review. As for #8257, since both PRs are touching the same functions in the same file it might be difficult to merge both change sets together.

darrenwee commented:
@thenaesh Ready for review. Thanks.

I think you can merge this one after review, then re-evaluate the other PR against the new `master`? Not sure what the correct protocol is.

damithc commented on src/main/java/teammates/common/util/StringHelper.java:
```diff
@@ -429,14 +411,12 @@ private static String handleNewLine(String str) {
-        return true;
+        return rowData.stream()
+                .anyMatch(r -> r.isEmpty());
     }
```
Have another look at this one. Doesn't look like the new code is same as previous.

thenaesh replied:
@darrenwee This should use `allMatch` and not `anyMatch`. If even a single element is nonempty this method returns false, so this method returns true exactly when all elements are empty.

@damithc Thanks for the catch, I missed this one out.

thenaesh replied:
Also, the fact that Travis passes means that we don't have a test for this...

darrenwee replied:
Oops, accidentally inverted this in my head. Yes, it should be the complement of `anyMatch`.

thenaesh commented on src/main/java/teammates/common/util/StringHelper.java:
```diff
@@ -312,34 +297,31 @@ public static String replaceIllegalChars(String str, String regex, char replacem
 
     public static String byteArrayToHexString(byte[] bytes) {
         StringBuilder sb = new StringBuilder(bytes.length * 2);
+
```
Extraneous change, please remove.

thenaesh commented on src/main/java/teammates/common/util/StringHelper.java:
```diff
@@ -312,34 +297,31 @@ public static String replaceIllegalChars(String str, String regex, char replacem
             }
             sb.append(Integer.toHexString(v));
         }
+
```
This too.

damithc commented:
For our peace of mind,  I suggest we add any missing tests in this PR too so that all methods touched in this PR are covered by tests. If the method is private, we can cover the public methods using that private method.

darrenwee commented:
Hey @thenaesh, this is ready for review again. I have added some test cases for some of the functions I modified, but the current test cases were pretty comprehensive. The missing test case for that `anyMatch` issue was resolved by another PR. I also fixed issues raised in your review.

There is one test case I added which I am unsure of. This (from `testJoinWithListOfIntegers`) passes:
```
assertEquals("5||14||null", StringHelper.join("||", Arrays.asList(5, 14, null)));
```
If you feed a `null` object to the list of integers, the resulting string from joining the list contains `null` as a string. I am not sure if this is desired behavior.

darrenwee commented:
@thenaesh good to go!

damithc commented on src/main/java/teammates/common/util/StringHelper.java:
```diff
@@ -87,11 +83,8 @@ public static String truncate(String inputString, int truncateLength) {
-        }
-        return result;
+
+        return inputString.substring(0, truncateLength - 3) + "...";
```
Does this work even if the inputString is shorter than `truncateLength`?

darrenwee replied:
Yes, but only if `inputString` is at least `truncateLength-3` characters long. That case is caught by the if-block before the return statement.

darrenwee commented:
Bump!

damithc commented:
@thenaesh for merging.
@darrenwee FYI, I've updated the branch (via GitHub UI) as it has gone out of date. i.e., your local is behind your fork now.
